### PR TITLE
Cache function in scalar model generate

### PR DESCRIPTION
**Pull request by blegat**
Before this PR:
```
julia> solve_scs_moi(data, vector_version = false, max_iters = 1)
-7.320596134944753e6

julia> TimerOutputs.get_defaulttimer()
 ─────────────────────────────────────────────────────────────────────────
                                  Time                   Allocations      
                          ──────────────────────   ───────────────────────
     Tot / % measured:         5.41s / 15.0%            322MiB / 100%     

 Section          ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────────
 SCS MOI scalar        1    808ms   100%   808ms    322MiB  100%    322MiB
   generate            1    566ms  70.1%   566ms    238MiB  73.9%   238MiB
   solve               1    242ms  29.9%   242ms   83.8MiB  26.0%  83.8MiB
 ─────────────────────────────────────────────────────────────────────────
```
After this PR:
```
julia> solve_scs_moi(data, vector_version = false, max_iters = 1)
-8.575051909370726e6

julia> TimerOutputs.get_defaulttimer()
 ─────────────────────────────────────────────────────────────────────────
                                  Time                   Allocations      
                          ──────────────────────   ───────────────────────
     Tot / % measured:         3.24s / 26.8%            305MiB / 100%     

 Section          ncalls     time   %tot     avg     alloc   %tot      avg
 ─────────────────────────────────────────────────────────────────────────
 SCS MOI scalar        1    870ms   100%   870ms    305MiB  100%    305MiB
   generate            1    580ms  66.6%   580ms    221MiB  72.5%   221MiB
   solve               1    290ms  33.4%   290ms   83.8MiB  27.5%  83.8MiB
 ─────────────────────────────────────────────────────────────────────────
```
The timings are not so stable but the allocations are. I'm surprised we don't gain more on allocations for `generate` by the way.

### CODE REVIEW

**Comment by odow** [comment]
This is kinda cheating. We don't want to win on these micro-optimizations.

**Comment by blegat** [comment]
The reason we need to copy every function coming in an `MOIU.Model` is precisely because we allow the user to do that.
Doing new copies both at the MOIU model level and at the user level doesn't seem right. No need to copy if the user creates new copies anyway.

**Comment by odow** [comment]
The issue is that this code is not standard code that the user or JuMP would write.

How much of a difference does it make on the full benchmark table?

**Comment by blegat** [comment]
Need to check the full benchmark.
Maybe JuMP should cache a function if it gives good perf improvements. It wouldn't hard to implement

File: pmedian.jl [hunk 1]
```diff
@@ -46,70 +46,54 @@ function generate_moi_problem(model, data::PMedianData)
...
-    MOI.add_constraint(
-        model,
-        MOI.ScalarAffineFunction(
-            MOI.ScalarAffineTerm.(1.0, facility_variables),
-            0.0,
-        ),
-        MOI.EqualTo{Float64}(data.num_facilities),
-    )
+    for i in 1:NC, j in 1:NL
+        func.terms[1] = MOI.ScalarAffineTerm(1.0, assignment_variables[i, j])
```
Comment (mlubin) [inline pmedian.jl:93]: The biggest gain is from avoiding allocating the small arrays here, right? How about avoiding the readability hit elsewhere and just re-using the function for this block?

Comment (blegat) [reply pmedian.jl:93]: Then it's not consistent and the reader of the code will wonder why not do this everywhere.

Comment (mlubin) [reply pmedian.jl:93]: Making the code uniformly more complex isn't a win for readability. Instead, we can add a comment here explaining why we reuse arrays.

**Comment by blegat** [comment]
With GLPK:
Before the PR:
```
julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          5.49s / 8.64%            270MiB / 76.2%    

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    475ms   100%   475ms    206MiB  100%    206MiB
   solve                1    418ms  88.1%   418ms    158MiB  76.6%   158MiB
   generate             1   55.8ms  11.7%  55.8ms   48.2MiB  23.4%  48.2MiB
 ──────────────────────────────────────────────────────────────────────────

julia> reset_timer!(TimerOutputs.get_defaulttimer());

julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          2.66s / 23.3%            206MiB / 100%     

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    620ms   100%   620ms    206MiB  100%    206MiB
   solve                1    558ms  90.1%   558ms    158MiB  76.6%   158MiB
   generate             1   60.7ms  9.80%  60.7ms   48.2MiB  23.4%  48.2MiB
 ──────────────────────────────────────────────────────────────────────────

julia> reset_timer!(TimerOutputs.get_defaulttimer());

julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          3.01s / 17.3%            206MiB / 100%     

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    520ms   100%   520ms    206MiB  100%    206MiB
   solve                1    449ms  86.4%   449ms    158MiB  76.6%   158MiB
   generate             1   70.4ms  13.5%  70.4ms   48.2MiB  23.4%  48.2MiB
 ──────────────────────────────────────────────────────────────────────────

julia> reset_timer!(TimerOutputs.get_defaulttimer());

julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          2.91s / 17.8%            206MiB / 100%     

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    517ms   100%   517ms    206MiB  100%    206MiB
   solve                1    473ms  91.3%   473ms    158MiB  76.6%   158MiB
   generate             1   44.1ms  8.53%  44.1ms   48.2MiB  23.4%  48.2MiB
 ──────────────────────────────────────────────────────────────────────────
```
After the PR:
```
julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          5.49s / 11.1%            254MiB / 74.6%    

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    612ms   100%   612ms    189MiB  100%    189MiB
   solve                1    507ms  82.9%   507ms    158MiB  83.4%   158MiB
   generate             1    104ms  17.0%   104ms   31.4MiB  16.6%  31.4MiB
 ──────────────────────────────────────────────────────────────────────────

julia> reset_timer!(TimerOutputs.get_defaulttimer());

julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          3.36s / 17.5%            189MiB / 100%     

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    588ms   100%   588ms    189MiB  100%    189MiB
   solve                1    550ms  93.5%   550ms    158MiB  83.4%   158MiB
   generate             1   37.8ms  6.43%  37.8ms   31.4MiB  16.6%  31.4MiB
 ──────────────────────────────────────────────────────────────────────────

julia> reset_timer!(TimerOutputs.get_defaulttimer());

julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
^[[A0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          2.95s / 17.7%            189MiB / 100%     

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    521ms   100%   521ms    189MiB  100%    189MiB
   solve                1    456ms  87.4%   456ms    158MiB  83.4%   158MiB
   generate             1   65.2ms  12.5%  65.2ms   31.4MiB  16.6%  31.4MiB
 ──────────────────────────────────────────────────────────────────────────

julia> reset_timer!(TimerOutputs.get_defaulttimer());

julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          2.65s / 20.9%            189MiB / 100%     

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    554ms   100%   554ms    189MiB  100%    189MiB
   solve                1    510ms  92.1%   510ms    158MiB  83.4%   158MiB
   generate             1   43.0ms  7.77%  43.0ms   31.4MiB  16.6%  31.4MiB
 ──────────────────────────────────────────────────────────────────────────

julia> reset_timer!(TimerOutputs.get_defaulttimer());

julia> solve_glpk_moi(data, vector_version = false, time_limit_sec = 0.001)
0.0

julia> TimerOutputs.get_defaulttimer()
 ──────────────────────────────────────────────────────────────────────────
                                   Time                   Allocations      
                           ──────────────────────   ───────────────────────
     Tot / % measured:          2.53s / 18.0%            189MiB / 100%     

 Section           ncalls     time   %tot     avg     alloc   %tot      avg
 ──────────────────────────────────────────────────────────────────────────
 GLPK MOI scalar        1    455ms   100%   455ms    189MiB  100%    189MiB
   solve                1    420ms  92.4%   420ms    158MiB  83.4%   158MiB
   generate             1   34.0ms  7.48%  34.0ms   31.4MiB  16.6%  31.4MiB
 ──────────────────────────────────────────────────────────────────────────
```

**Comment by odow** [comment]
What is the difference in the latex table?

**Comment by odow** [comment]
What `data` did you use? When I ran this, I got essentially no change in the full latex table. I'm in favor of closing.

**Comment by odow** [comment]
This doesn't seem like it is happening.

**Pull request closed without merging**